## 1. How many Spam Emails are there?

In [1]:
def count_spam(path = "C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data", folders = ['enron1','enron2']):
    import os
    from pathlib import Path
    path = Path(path)
    spam_count = 0
    for folder in folders:
        file_names = os.listdir(path / folder / 'spam')
        for file_name in file_names:
            if 'spam' in file_name:
                spam_count += 1
    print('There are', spam_count, 'spam emails in', folders)

In [2]:
count_spam()

There are 2993 spam emails in ['enron1', 'enron2']


## 2. Structure the email data from the 2 directories into 1 dataframe with columns: Status, Subject, Body.

#### Specify the folders that will be included and the path leading to the folders.

In [3]:
folders = ['enron1','enron2']
status = ['spam','ham']
path1 = "C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\{folder}\\{status}"

#### Create the paths containing all the desired emails.

In [4]:
from itertools import product
directories = list(product(folders, status))
directories

[('enron1', 'spam'), ('enron1', 'ham'), ('enron2', 'spam'), ('enron2', 'ham')]

In [5]:
paths = []
for i in directories:
    paths.append(path1.format(folder=i[0],status=i[1]))
paths

['C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\spam',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron2\\ham']

#### List all the files in each `path`.

In [6]:
import glob
file_paths = [glob.glob(pathname=path + '\\*') for path in paths]

#### Now we have all the file paths of the emails in the specified folders.

In [7]:
file_paths[0][0:5]

['C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0006.2003-12-18.GP.spam.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0008.2003-12-18.GP.spam.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0017.2003-12-18.GP.spam.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0018.2003-12-18.GP.spam.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\spam\\0026.2003-12-18.GP.spam.txt']

In [8]:
file_paths[1][0:5]

['C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0001.1999-12-10.farmer.ham.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0002.1999-12-13.farmer.ham.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0003.1999-12-14.farmer.ham.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0004.1999-12-14.farmer.ham.txt',
 'C:\\Users\\jakeedison.bersabe\\Cadetship\\EDA\\apper-eda\\Notebooks\\data\\enron-data\\enron1\\ham\\0005.1999-12-14.farmer.ham.txt']

#### Having acquired the file paths, we can now extract the status, subject and message body of the emails.

In [9]:
subjects = []
bodies = []
status = []

# for each path
for file_path in file_paths:
    # for each file in path
    for i in file_path:
        # for status
        if 'spam.txt' in i:
            status.append('spam')
        elif 'ham.txt' in i:
            status.append('ham')
        # extract the subject from email
        with open(i,'r',errors='ignore') as f:
            subject = f.readlines()[0]
            subject = subject[9:]
            subjects.append(subject)
        # extract body message from email
        with open(i,'r',errors='ignore') as f:
            body = f.readlines()
            body = body[1:]
            bodies.append(body)
            
# join each sub-elements of the bodies list
bodies = [' '.join(i) for i in bodies]

print('Status has', len(status), 'elements.')
print('Subject has', len(subjects), 'elements.')
print('Body has', len(bodies), 'elements.')

Status has 11026 elements.
Subject has 11026 elements.
Body has 11026 elements.


#### Now, use the lists to create a dataframe.

In [10]:
import pandas as pd

In [11]:
enron = pd.DataFrame(list(zip(status,subjects,bodies)), columns=['Status','Subject','Body'])
enron.sample(10)

Status                                            Subject  \
825     spam                   i promised to help 5000 people\n   
1698     ham  is this fri feb 11 a problem for taking vacati...   
8465     ham                   re : rabi de ' s sign on bonus\n   
7780     ham             pricing credit on thousands of names\n   
10604    ham                                        re : test\n   
7056     ham              re : summer intern : paulo oliveira\n   
9787     ham                     vince kaminski bullet points\n   
999     spam  get laid tonight : do you need someone to touch\n   
1827     ham                         cdnow order confirmation\n   
5753    spam  we have been rated as # 1 one - stop - shop in...   

                                                    Body  
825    please refer all questions , opinions or addit...  
1698                                                      
8465   sorry vince i have been out of town . please g...  
7780   all -\n our challenge for the next few months ...  
10604  dear vince : the email address of candice is\n...  
7056   vince , both matt and april think that this ty...  
9787   * definition of price volatility\n * historica...  
999    goodbye\n crevice evade inviablesplice\n phosp...  
1827   dear daren ,\n thank you for shopping at cdnow...  
5753   take the pill and enjoy great sex\n commit a c...

## 3. Build a Naive Bayes classifier to classify whether emails are spam or not.

#### Import libraries

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [13]:
enron.sample(5)

Status                                            Subject  \
584    spam                                         for 9 or\n   
8724    ham                          re : from vicky windsor\n   
5624   spam                  in the heart of your business !\n   
4395    ham  revision : eastrans nomination change effectiv...   
4372    ham                                      re : cp & l\n   

                                                   Body  
584   email loading . . . %\n wise triatomic lauren ...  
8724  vicky ,\n please , send me your resume .\n i s...  
5624  corporate image can say a lot of things about ...  
4395  deliveries and redeliveries will still be 0 ef...  
4372  daren ,\n i am handling cp & l and am trying t...

In [14]:
enron.shape

(11026, 3)

#### Create `Spam` column from the `Status` column and delete `Status` column.

In [15]:
enron['Spam'] = [1 if i == 'spam' else 0 for i in enron['Status']]
enron.drop('Status', inplace=True, axis=1)

In [16]:
enron.sample(5)

Subject  \
8679                        re : subscription renewals\n   
6307   hot stock info : drgv announces another press ...   
8961                             re : enron case study\n   
10817                              re : corporate card\n   
9121                                 pserc iab meeting\n   

                                                    Body  Spam  
8679   susan :\n vince would like to renew all of the...     0  
6307   a $ 3 , 800 investment could be worth $ 50 , 0...     1  
8961   outstanding , cindy . thank you so much . i wi...     0  
10817  yes tony ,\n mike authorized you for a corp . ...     0  
9121   dear dennis ,\n vince kaminski forwarded your ...     0

#### Split data for training and testing.

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
train_X, test_X, train_y, test_y = train_test_split(enron["Body"], enron["Spam"], test_size=0.3, random_state=0)

#### Import libraries for text analysis

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import GaussianNB

In [20]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

#### Train the model

In [21]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

## 4. What is the longest ham email?

#### Create a dataframe containing all ham emails.

In [22]:
hams = enron[enron['Spam'] == 0][['Body']]

#### Create a column containing the lengths of each emails.

In [23]:
hams['Lengths'] = [len(i) for i in hams['Body']]
hams.drop('Body',inplace=True,axis=1)

#### Get the index of the longest ham email.

In [24]:
longest_ham = hams.idxmax()
print(longest_ham)

Lengths    10765
dtype: int64


#### Print the longest ham email.

In [25]:
enron.iloc[longest_ham,1]

10765    fyi news articles from indian press .\n - - - ...
Name: Body, dtype: object

#### The longest ham email has `43888` characters.

In [26]:
hams.max()

Lengths    43888
dtype: int64

## 5. What is the accuracy of your model?

In [27]:
from sklearn.model_selection import cross_val_score

In [28]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(3308, 58021)
Accuracy: 94.01 percent


**The model has an accuracy of `94.01%`.**

## 6. Include the Subject in the analysis of the emails, does the accuracy/performance of the model increase?

#### Create a new column containing the concatenation of `Subject` and `Body`.

In [29]:
enron['Total'] = [sub + bod for (sub, bod) in zip(enron.iloc[:,0],enron.iloc[:,1])]

In [30]:
enron.head(5)

Subject  \
0  dobmeos with hgh my energy level has gone up !...   
1        your prescription is ready . . oxwq s f e\n   
2                            get that new car 8434\n   
3                              await your response\n   
4  coca cola , mbna america , nascar partner with...   

                                                Body  Spam  \
0  introducing\n doctor - formulated\n hgh\n huma...     1   
1  low cost prescription medications\n soma , ult...     1   
2  people nowthe weather or climate in any partic...     1   
3  dear partner ,\n we are a team of government o...     1   
4  stock\n profile\n about\n company\n investment...     1   

                                               Total  
0  dobmeos with hgh my energy level has gone up !...  
1  your prescription is ready . . oxwq s f e\nlow...  
2  get that new car 8434\npeople nowthe weather o...  
3  await your response\ndear partner ,\n we are a...  
4  coca cola , mbna america , nascar partner with...

In [31]:
train_X, test_X, train_y, test_y = train_test_split(enron["Total"], enron["Spam"], test_size=0.3, random_state=0)

In [32]:
vectorizer = TfidfVectorizer(stop_words='english')
tfidf_train_X = vectorizer.fit_transform(train_X)

In [33]:
classifier = GaussianNB()
classifier.fit(tfidf_train_X.toarray(), train_y)

GaussianNB()

In [34]:
tfidf_test_X = vectorizer.transform(test_X)
print(tfidf_test_X.shape)
scores = cross_val_score(classifier, tfidf_test_X.toarray(), test_y, cv=5)
acc = scores.mean()
print("Accuracy: %0.2f percent" % (acc *100))

(3308, 58864)
Accuracy: 94.68 percent


**The accuracy improved by `0.67%`**

## Bonus : Answer questions 1-6 using Enron 1,2, and 3.

**Answer to bonus question is in the jupyter notebook `PS3 - Bonus - Jake.ipynb`.**